### ISSUE TITLE
How to handle Nuget packages referenced from repositories other than Nuget.org

### ISSUE THREAD

**Issue by AlexandreRoba** [projectkudu/kudu#1513]
We deploying our sites we are make use of some NuGet packages that are stored on a different repository than the one at Nuget.org. The Packages from the Nuget.org are installed without any issue but when we reach the NuGets from the different repository we are getting the following error on the " Running deployment command... " action:
Command: "C:\inetpub\apps\helloidentity\site\deployments\tools\deploy.cmd"
Handling .NET Web Application deployment.
Unable to find version '3.0.20820.0' of package 'AjaxControlToolkit'.
Unable to find version '2.1.0.0' of package 'DirectShowLib-2005'.
Unable to find version '2.1.9.0' of package 'linqtotwitterNET40'.
Unable to find version '1.0.0.0' of package 'Microsoft.Http'.
Unable to find version '15.0.4569.1000' of package 'Microsoft.SharePoint.Client'.
Unable to find version '6.6.6' of package 'MySql.Data'.
Unable to find version '4.3.1.0' of package 'payflow_dotNET'.
Unable to find version '3.0.8.0' of package 'Sitefinity.RazorEngine'.
Failed exitCode=1, command=nuget restore "C:\inetpub\apps\helloidentity\site\repository\src\HelloIdentityWebSite\HelloIdentityWebSite.sln"
Unable to find version '0.5.1.6' of package 'Sitefinity.RedDog.Search'.
An error has occurred during web site deployment.

We have enabled the package restore on the solution and the repository is defined as a package source on nuget.target file which is under source control under the .nyget folder at the solution level. When building on TFS we are not getting any error and the packages gets restored properly using the repository defined on the NuGet.target.

I tried to add a NuGet.Config on the same folder as the NuGet.exe which is used by kudu "C:\inetpub\Kudu.Services.Web\bin\Scripts\Nuget.Config" with the following:
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="Nuget" value="https://www.nuget.org/api/v2/" />
    <add key="SiteFinity" value="http://http://nuget.sitefinity.com/nuget/" />
  </packageSources>
</configuration>
```
But it still cannot find the Nugets... Any idea how I can resolve this?

**Comment by pranavkm** [comment]
Are you sure the `SiteFinity` source is specified correctly in your `NuGet.config`? It has two `http://` in the url

**Comment by AlexandreRoba** [comment]
Indeed you are right. Unfortunatlly this does not resolve might issue...The typo was comming from my trial to add a NuGet.config file on the scripts folder. Apparently this config file is not picked up :(
I checked my Nuget.target file on the solution and it does contains the right values...

<ItemGroup Condition=" '$(PackageSources)' == '' ">
      <PackageSource Include="https://www.nuget.org/api/v2/" />
      <PackageSource Include="http://nuget.sitefinity.com/nuget/" />
    </ItemGroup>

**Comment by AlexandreRoba** [comment]
I looked at the deployment script "deploy.cmd".

It calls the following command:
:: 1. Restore NuGet packages
IF /I "src\HelloIdentityWebSite\HelloIdentityWebSite.sln" NEQ "" (
  call :ExecuteCmd nuget restore "%DEPLOYMENT_SOURCE%\src\HelloIdentityWebSite\HelloIdentityWebSite.sln"
  IF !ERRORLEVEL! NEQ 0 goto error
)

I tried to run the nuget restore command manualy from the repository folder:

C:\inetpub\apps\helloidentity\site\repository [master]> nuget restore .\src\HelloIdentityWebSite\HelloIdentityWebSite.sl
n
Installing 'MySql.Data 6.6.6'.
Installing 'linqtotwitterNET40 2.1.9.0'.
Installing 'Microsoft.Http 1.0.0.0'.
Installing 'DirectShowLib-2005 2.1.0.0'.
Installing 'payflow_dotNET 4.3.1.0'.
Installing 'Sitefinity.RazorEngine 3.0.8.0'.
Installing 'Microsoft.SharePoint.Client 15.0.4569.1000'.
Installing 'AjaxControlToolkit 3.0.20820.0'.
Successfully installed 'Microsoft.Http 1.0.0.0'.
Successfully installed 'payflow_dotNET 4.3.1.0'.
Successfully installed 'DirectShowLib-2005 2.1.0.0'.
Successfully installed 'linqtotwitterNET40 2.1.9.0'.
Successfully installed 'Microsoft.SharePoint.Client 15.0.4569.1000'.
Successfully installed 'AjaxControlToolkit 3.0.20820.0'.
Successfully installed 'Sitefinity.RazorEngine 3.0.8.0'.
Successfully installed 'MySql.Data 6.6.6'.
Installing 'Sitefinity.Windows.Azure 2.0.0.1'.
Installing 'Sitefinity.RedDog.Search 0.5.1.6'.
Successfully installed 'Sitefinity.RedDog.Search 0.5.1.6'.
Installing 'System.Windows 5.0.61118.0'.
Installing 'Telerik.Sitefinity.Analytics 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Audit.Elasticsearch 8.0.5701.0'.
Installing 'Telerik.Sitefinity.All 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.All 8.0.5701.0'.
Installing 'Telerik.Sitefinity.AuditTrail 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Connectivity 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Audit.Elasticsearch 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.AuditTrail 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Connectivity 8.0.5701.0'.
Successfully installed 'System.Windows 5.0.61118.0'.
Successfully installed 'Sitefinity.Windows.Azure 2.0.0.1'.
Successfully installed 'Telerik.Sitefinity.Analytics 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Content 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Core 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Dashboard 8.0.5701.0'.
Installing 'Telerik.Sitefinity.DAM 8.0.5701.0'.
Installing 'Telerik.Sitefinity.DigitalExperienceCloudConnector 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Diagnostics 8.0.5701.0'.
Installing 'Telerik.Sitefinity.DropboxLibraries 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Everlive 8.0.5701.0'.
Installing 'Telerik.Sitefinity.EverliveConnector 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.EverliveConnector 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.DigitalExperienceCloudConnector 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Dashboard 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.DropboxLibraries 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Diagnostics 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.DAM 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Everlive 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Feather 1.2.120.0'.
Installing 'Telerik.Sitefinity.Feather.Core 1.2.120.0'.
Installing 'Telerik.Sitefinity.MarketoConnector 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Octopus 2015.1.225.1'.
Installing 'Telerik.Sitefinity.Migration 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Mvc 8.0.5701.0'.
Installing 'Telerik.Sitefinity.MobileAppBuilder 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.MarketoConnector 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Mvc 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.MobileAppBuilder 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Migration 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Feather.Core 1.2.120.0'.
Successfully installed 'Telerik.Sitefinity.Content 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Octopus 2015.1.225.1'.
Installing 'Telerik.Sitefinity.PrecompiledPages.Backend 8.0.5701.0'.
Installing 'Telerik.Sitefinity.OpenAccess 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Reporting.Controls 7.1.13.612'.
Installing 'Telerik.Sitefinity.PrecompiledTemplates 8.0.5701.0'.
Installing 'Telerik.Sitefinity.RecycleBin 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.PrecompiledPages.Backend 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.OpenAccess 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.RecycleBin 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Reporting.Controls 7.1.13.612'.
Installing 'Telerik.Sitefinity.SalesForceConnector 8.0.5701.0'.
Installing 'Telerik.Sitefinity.ServicesImpl 8.0.5701.0'.
Installing 'Telerik.Sitefinity.SharepointConnector 8.0.5701.0'.
Installing 'Telerik.Sitefinity.SiteSync 8.0.5701.0'.
Installing 'Telerik.Sitefinity.SitemapGenerator 8.0.5701.0'.
Installing 'Telerik.Sitefinity.TfisClient 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.SitemapGenerator 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.SalesForceConnector 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.SharepointConnector 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.SiteSync 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.TfisClient 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.PrecompiledTemplates 8.0.5701.0'.
Installing 'Telerik.Web.UI 2015.1.225.40'.
Installing 'Telerik.Sitefinity.Windows.Controls 2013.1.415.40'.
Installing 'Telerik.Sitefinity.Thunder 8.0.5701.0'.
Installing 'Telerik.Sitefinity.Twitterizer 2.4.1'.
Successfully installed 'Telerik.Sitefinity.ServicesImpl 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Twitterizer 2.4.1'.
Successfully installed 'Telerik.Sitefinity.Thunder 8.0.5701.0'.
Successfully installed 'Telerik.Sitefinity.Windows.Controls 2013.1.415.40'.
Successfully installed 'Telerik.Sitefinity.Feather 1.2.120.0'.
Successfully installed 'Telerik.Sitefinity.Core 8.0.5701.0'.
Successfully installed 'Telerik.Web.UI 2015.1.225.40'.
All packages listed in packages.config are already installed.
C:\inetpub\apps\helloidentity\site\repository [master]>

There it restore the package without any issue.
Could it be that the LocalSystem which is used to run the deploy.cmd from the application pool cannot access the site repository?....

**Comment by davidebbo** [comment]
If you click Retry for the deployment, does it fail consistently? That runs in the same context as when using the Kudu console, so it should work.

**Comment by AlexandreRoba** [comment]
If I run the nuget restore  manually from the repository folder it downloads the package into the packages folder and then running the Deploy again proceed (but then it fails again for another reason that I'm investigating I might create another issue for this one. :)) 
If I run a Redeploy(clean) then if fails but it make sense as It removes the nuget from the packages folder and try to redownload them.. I still have no idea why it is not downloading the packages from the SiteFinity Repository :(

**Comment by davidebbo** [comment]
Could you [share a repro repository](https://github.com/projectkudu/kudu/wiki/Using-a-git-repo-to-report-an-issue) so we can take a closer look? And if you create a new test site and deploy there, does it happen consistently?

**Comment by AlexandreRoba** [comment]
Thanks for the help. I have compiled a dummy repo that reproduce the problem.
It is a dummy ASP.NET mvc application. I could deploy it without any issue when all packages came from the Nuget.org repo. I then have added the package AjaxControlToolkit fromt the Telerik Sifinity Repo and then I get the error the package cannot be restored...

Command: "C:\inetpub\apps\mvcapplication\site\deployments\tools\deploy.cmd"
Handling .NET Web Application deployment.
Unable to find version '3.0.20820.0' of package 'AjaxControlToolkit'.
Failed exitCode=1, command=nuget restore "C:\inetpub\apps\mvcapplication\site\repository\src\MVCApplicationTest\MVCApplicationTest.sln"
An error has occurred during web site deployment.


The repo i'm trying to deploy can be found on github at:
https://github.com/AlexandreRoba/MVCApplicationTest

thanks for any help you can provide me on this :)

**Comment by davidebbo** [comment]
This version indeed doesn't exist on http://www.nuget.org/packages/AjaxControlToolkit/. It doesn't seem like you're specifying an alternate feed anywhere, so the error is expected. What am I missing here?

BTW, you should not be enabling NuGet package restore. That was the old way of doing it. See http://blog.davidebbo.com/2014/01/the-right-way-to-restore-nuget-packages.html. You should not  have nuget.exe in your repo at all.

**Comment by AlexandreRoba** [comment]
I'm specifying the alternate place of the repository on the nuget.targets file which is created by the "Enable Nuget restore". You can find it there: https://github.com/AlexandreRoba/MVCApplicationTest/blob/master/src/MVCApplicationTest/.nuget/NuGet.targets

I'll check the "new way" of restoring packages and I'll try that. I'll keep you posted.

**Comment by davidebbo** [comment]
Custom feeds are set using nuget.config, which you can have next to your sln file. See [here](https://docs.nuget.org/consume/nuget-config-settings)

**Comment by AlexandreRoba** [comment]
I made the correction. I check Xavier Decoster Blog that raised also couple of issues with the "old" way of enabling package restore aka MSBuild-Integrated Package Restore.

In my case I was not able to enable package restore this way with kudu. This approach is still supported by NuGet and I totally understand that an open source project as other things to do than supporting a becoming deprecated functionality of NuGet. :)

I have enabled the Automatic package restore since and it works like a charm.

Thanks for your support.
